### PR TITLE
Introduce getSynthesizedDeepClone

### PR DESCRIPTION
**Pull request by amcasey**
During refactoring, it seems quite common to require that a node appear in more than one place in the rewritten tree.

### CODE REVIEW

**Comment by amcasey** [comment]
Apparently, this is a public API change.  What's the correct way to handle that?

**Comment by weswigham** [comment]
If it's desirable, just accept the new API baselines.

**Comment by andy-ms** [comment]
Or mark it `@internal` like `getSynthesizedClone`.

**Comment by amcasey** [comment]
@andy-ms's right - it was an oversight on my part.

File: src/compiler/factory.ts [hunk 1]
```diff
@@ -66,6 +71,16 @@ namespace ts {
         return clone;
     }
 
+    /**
+     * Creates a deep, memberwise clone of a node with no source map location.
+     */
+    /* @internal */
+    export function getSynthesizedDeepClone<T extends Node>(node: T | undefined): T | undefined {
+        return node
+            ? getSynthesizedClone(visitEachChild(node, child => getSynthesizedDeepClone(child), nullTransformationContext))
```
Comment (rbuckton) [inline src/compiler/factory.ts]: No need to write `child => getSynthesizedDeepClone(child)` when just `getSynthesizedDeepClone` would do.

File: src/compiler/factory.ts [hunk 2]
```diff
@@ -66,6 +71,16 @@ namespace ts {
         return clone;
     }
 
+    /**
+     * Creates a deep, memberwise clone of a node with no source map location.
+     */
+    /* @internal */
+    export function getSynthesizedDeepClone<T extends Node>(node: T | undefined): T | undefined {
+        return node
```
Comment (rbuckton) [inline src/compiler/factory.ts]: This whole operation seems somewhat expensive. The call to `visitEachChild` will likely resolve in a call to one of the `updateX` functions in _factory.ts_, which in turn calls `updateNode`, which in turn sets the `original` pointer on the new node, clones the original node's `emitNode`, and sets the `pos` and `end` of the new node. Then `getSynthesizedClone` discards the new object for a fresh one.

If we only need this for entity-name-like expressions, could we instead write a more specific version that doesn't do all of this extra work that is just thrown out? See `serializeQualifiedNameAsExpression` in _ts.ts_ for a function that already does this (albeit with some extra options you don't need).

Comment (amcasey) [reply src/compiler/factory.ts]: I agree that's it's overkill for this application.  It was actually added for https://github.com/Microsoft/TypeScript/pull/18997 and this was a proof-of-concept application.  If there's a way to make this less expensive, I'd be happy to rework it.

Comment (rbuckton) [reply src/compiler/factory.ts]: Also, consider moving this to _services_ so that it isn't used unintentionally in _compiler_.

**Comment by amcasey** [comment]
@rbuckton I believe I've addressed your feedback.

**Comment by amcasey** [comment]
@rbuckton - I believe most/all callers will want to copy trivia. Does that mean deep-clone should actually be built on getMutableClone, rather than getSynthesizedClone?

**Comment by rbuckton** [comment]
I'd rather not use `getMutableClone` as I'd like to eventually get rid of that function. You could always call `setTextRange` in your `getDeepSynthesizedClone` to copy positions.

**Comment by amcasey** [comment]
@rbuckton Could I just refrain from clearing `pos` and `end`?  (I assume I still have to clear `parent`.)

**Comment by amcasey** [comment]
As of this PR, the positions don't make a difference.  I'll add a test once https://github.com/Microsoft/TypeScript/pull/18931 is merged.

**Comment by amcasey** [comment]
@rbuckton signed off offline.

**Pull request merged**